### ISSUE TITLE
[Feature Request] In-workflow timer trigger based activities

### ISSUE THREAD

**Issue by SimonLuckenuik** [Azure/azure-functions-durable-extension#174]
**Scenario Example**
We are looking into using Durable Functions for a workflow that requires a lock mechanism:
- Take a lease on a blob and renew it every 30 seconds, while the workflow is running. 

**Implementation with current framework**
Option 1:
1) Create timers that you renew everytime it times out. Always run those timers in parallel with the main workflow activities (need to wait on them) + manage the CancellationTokenSource
2) When timer expires, start a lease renew activity (could be fire and forget)

Option 2:
1) Start a sub-orchestration which consists of only a timer that expires and renew the lock.

**Current implementation problem**
Option 1:
1) The code written to support that scenario is heavy and hard to follow and maintain
2) Cannot rely on a Timer trigger normal function, because the workflow is not running frequently and every execution of the function while the workflow is not running would be a waste of money. Would need to keep a list of active workflows where to renew lock. Not convenient,

Option 2:
1) looking at the documentation, it is not clear what happen to the sub-orchestration once the main orchestration terminates (success or failure).

**Suggestion**
- Introduce concept of "in-workflow timer trigger", that would trigger an activity at a specific interval (like a cron job), while the workflow is running only.

**Comment by cgillum** [comment]
Can you give an example of what this would look like in code?

**Comment by SimonLuckenuik** [comment]
Off the top of my head, what I am suggesting could be something like that to make it simple and close to the TimerTrigger:
```
[FunctionName("RenewLease")]
public static async Task RenewLease(
[ActivityTrigger] DurableActivityContext ctx,         
[WorkflowTimerTrigger("0 */5 * * * *")] TimerInfo timer)
{
    string leaseToRenew = ctx.GetInput<string>();
    await RenewLease(leaseToRenew);
}
```

**Comment by cgillum** [comment]
I don't think it is likely we'll be able to implement this exact feature given the narrow use-case. However, in the next release we are introducing support for Azure Event Grid publishing: https://github.com/Azure/azure-functions-durable-extension/issues/216. I can imagine that you could use this to start and stop background orchestrations for the instances you care about.

Would something like that work for you?

**Comment by SimonLuckenuik** [comment]
I currently can implement what I am looking for, this issue was mainly to have nicer and simpler code. 

Are there any plans for "Named Singleton" concept? That would do the trick as well to manage locking scenarios\resource exclusivity (a bit like we would do with Named Mutex in conventionnal threading).

**Comment by cgillum** [comment]
Yes, we support named singletons today. Documentation here: https://docs.microsoft.com/en-us/azure/azure-functions/durable-functions-singletons.

**Comment by cgillum** [comment]
Fair point. That's a good reason for us to get `[Singleton]` officially supported in consumption. Others have worked around this using blob leases, but ideally we have something that doesn't require so much work.
